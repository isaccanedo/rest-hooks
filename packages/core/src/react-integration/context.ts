import { createContext } from 'react';
import type { DenormalizeCache } from '@rest-hooks/normalizr';

import type { ActionTypes, State } from '../types.js';
import { initialState } from '../state/createReducer.js';
import Controller from '../controller/Controller.js';

export const StateContext = createContext(initialState);

const dispatch = (value: ActionTypes) => {
  /* istanbul ignore next */
  if (process.env.NODE_ENV !== 'production') {
    console.error(
      'It appears you are trying to use Rest Hooks without a provider.\nFollow instructions: https://resthooks.io/docs/getting-started/installation#add-provider-at-top-level-component',
    );
    /* istanbul ignore next */
    if (process.env.NODE_ENV === 'test') {
      console.error(
        'If you are trying to test: https://resthooks.io/docs/guides/unit-testing-hooks',
      );
    }
  }
  return Promise.resolve();
};
export const DispatchContext = createContext(dispatch);

// this is not needed anymore, but keeping around for backcompatibility
export const DenormalizeCacheContext = createContext<DenormalizeCache>({
  entities: {},
  results: {},
});

export const ControllerContext = createContext<Controller>(
  new Controller({
    dispatch,
    globalCache: {
      entities: {},
      results: {},
    },
  }),
);

export interface Store<S> {
  subscribe(listener: () => void): () => void;
  dispatch: React.Dispatch<ActionTypes>;
  getState(): S;
  uninitialized?: boolean;
}
/* istanbul ignore next */
export const StoreContext = createContext<Store<State<unknown>>>({
  subscribe: listener => () => {},
  dispatch: () => {},
  getState: () => initialState,
  uninitialized: true,
});
