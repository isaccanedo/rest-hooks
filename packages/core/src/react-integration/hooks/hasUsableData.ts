import { FetchShape } from '../../endpoint/index.js';

/** @deprecated use https://resthooks.io/docs/api/Controller#getResponse */
export default function hasUsableData(
  fetchShape: Pick<FetchShape<any>, 'options'>,
  cacheReady: boolean,
  deleted: boolean,
  invalidated?: boolean,
) {
  return (
    !deleted &&
    !fetchShape.options?.invalidIfStale &&
    cacheReady &&
    !invalidated
  );
}
