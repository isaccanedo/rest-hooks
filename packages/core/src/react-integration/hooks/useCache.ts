/* eslint-disable @typescript-eslint/ban-types */
import { useMemo } from 'react';
import {
  EndpointInterface,
  DenormalizeNullable,
  Schema,
  FetchFunction,
  ResolveType,
} from '@rest-hooks/normalizr';

import { ReadShape, ParamsFromShape } from '../../endpoint/index.js';
import shapeToEndpoint from '../../endpoint/adapter.js';
import useCacheNew from '../newhooks/useCache.js';

/**
 * Access a response if it is available.
 *
 * `useCache` guarantees referential equality globally.
 * @see https://resthooks.io/docs/api/useCache
 */
export default function useCache<
  E extends
    | Pick<
        EndpointInterface<FetchFunction, Schema | undefined, undefined>,
        'key' | 'schema' | 'invalidIfStale'
      >
    | Pick<ReadShape<any, any>, 'getFetchKey' | 'schema' | 'options'>,
  Args extends
    | (E extends { key: any }
        ? readonly [...Parameters<E['key']>]
        : readonly [ParamsFromShape<E>])
    | readonly [null],
>(
  endpoint: E,
  ...args: Args
): E['schema'] extends {}
  ? DenormalizeNullable<E['schema']>
  : E extends (...args: any) => any
  ? ResolveType<E> | undefined
  : any {
  const adaptedEndpoint: any = useMemo(() => {
    return shapeToEndpoint(endpoint);
    // we currently don't support shape changes
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return useCacheNew(adaptedEndpoint, ...args);
}
