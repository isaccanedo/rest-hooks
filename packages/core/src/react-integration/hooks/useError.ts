import type { NetworkError, UnknownError } from '@rest-hooks/normalizr';
import { useMemo } from 'react';
import {
  EndpointInterface,
  Schema,
  FetchFunction,
} from '@rest-hooks/normalizr';

import { ReadShape, ParamsFromShape } from '../../endpoint/index.js';
import shapeToEndpoint from '../../endpoint/adapter.js';
import useErrorNew from '../newhooks/useError.js';

export type ErrorTypes = NetworkError | UnknownError;

type UseErrorReturn<P> = P extends [null] ? undefined : ErrorTypes | undefined;

/**
 * Get any errors for a given request
 * @see https://resthooks.io/docs/api/useError
 */
export default function useError<
  E extends
    | Pick<
        EndpointInterface<FetchFunction, Schema | undefined, undefined>,
        'key' | 'schema' | 'invalidIfStale'
      >
    | Pick<ReadShape<any, any>, 'getFetchKey' | 'schema' | 'options'>,
  Args extends
    | (E extends { key: any }
        ? readonly [...Parameters<E['key']>]
        : readonly [ParamsFromShape<E>])
    | readonly [null],
>(endpoint: E, ...args: Args): UseErrorReturn<typeof args> {
  const adaptedEndpoint: any = useMemo(() => {
    return shapeToEndpoint(endpoint);
    // we currently don't support shape changes
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return useErrorNew(adaptedEndpoint, ...args);
}
