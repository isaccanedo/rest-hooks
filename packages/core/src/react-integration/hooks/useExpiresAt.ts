import { ReadShape, ParamsFromShape } from '../../endpoint/index.js';
import useMeta from './useMeta.js';

/** Returns whether the data at this url is fresh or stale
 * @deprecated use https://resthooks.io/docs/api/Controller#getResponse
 */
export default function useExpiresAt<
  Shape extends Pick<ReadShape<any, any>, 'getFetchKey' | 'options'>,
>(
  fetchShape: Shape,
  params: ParamsFromShape<Shape> | null,
  entitiesExpireAt = 0,
): number {
  const meta = useMeta(fetchShape, params);

  if (!meta) {
    return entitiesExpireAt;
  }

  return meta.expiresAt;
}
