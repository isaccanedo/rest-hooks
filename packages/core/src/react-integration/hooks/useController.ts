import { useContext } from 'react';

import Controller from '../../controller/Controller.js';
import { ControllerContext } from '../context.js';

/**
 * Imperative control of Rest Hooks store
 * @see https://resthooks.io/docs/api/useController
 */
export default function useController(): Controller {
  return useContext(ControllerContext);
}
