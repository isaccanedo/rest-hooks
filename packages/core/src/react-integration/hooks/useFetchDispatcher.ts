import { Schema } from '@rest-hooks/normalizr';
import { useContext, useCallback } from 'react';

import {
  FetchShape,
  SchemaFromShape,
  ParamsFromShape,
  BodyFromShape,
  OptimisticUpdateParams,
  ReturnFromShape,
} from '../../endpoint/index.js';
import { DispatchContext } from '../context.js';
import createFetch from '../../state/actions/createFetch.js';

/** Build an imperative dispatcher to issue network requests.
 * @deprecated use https://resthooks.io/docs/api/Controller#fetch
 */
export default function useFetchDispatcher(throttle = false): <
  Shape extends FetchShape<Schema, Readonly<object>, any>,
  UpdateParams extends OptimisticUpdateParams<
    SchemaFromShape<Shape>,
    FetchShape<any, any, any>
  >[],
>(
  fetchShape: Shape & {
    update?: (...args: any) => Record<string, (...args: any) => any>;
  },
  params: ParamsFromShape<Shape>,
  body: BodyFromShape<Shape>,
  updateParams?: UpdateParams | undefined,
) => ReturnFromShape<typeof fetchShape> {
  const dispatch = useContext(DispatchContext);

  const fetchDispatcher = useCallback(
    <Shape extends FetchShape<Schema, Readonly<object>, any>>(
      fetchShape: Shape,
      params: ParamsFromShape<Shape>,
      body: BodyFromShape<Shape>,
      updateParams?:
        | OptimisticUpdateParams<
            SchemaFromShape<Shape>,
            FetchShape<any, any, any>
          >[]
        | undefined,
    ) => {
      const action = createFetch(fetchShape, {
        params,
        body,
        throttle,
        updateParams,
      });
      dispatch(action);
      return action.meta.promise;
    },
    [dispatch, throttle],
  );
  // any is due to the ternary that we don't want to deal with in our implementation
  return fetchDispatcher as any;
}
