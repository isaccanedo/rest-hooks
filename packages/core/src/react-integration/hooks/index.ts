import useFetcher from './useFetcher.js';
import useController from './useController.js';
import useCache from './useCache.js';
import useRetrieve from './useRetrieve.js';
import useResource from './useResource.js';
import useSubscription from './useSubscription.js';
import useMeta from './useMeta.js';
import useError from './useError.js';
import useExpiresAt from './useExpiresAt.js';
import useInvalidator from './useInvalidator.js';
import useResetter from './useResetter.js';
import useFetchDispatcher from './useFetchDispatcher.js';
import useInvalidateDispatcher from './useInvalidateDispatcher.js';
import useFetch from '../newhooks/useFetch.js';
import useSuspense from '../newhooks/useSuspense.js';
export { default as hasUsableData } from './hasUsableData.js';
export type { ErrorTypes } from './useError.js';

export {
  useFetcher,
  useController,
  useFetchDispatcher,
  useCache,
  useError,
  useRetrieve,
  useResource,
  useSubscription,
  useMeta,
  useExpiresAt,
  useInvalidator,
  useInvalidateDispatcher,
  useResetter,
  useFetch,
  useSuspense,
};
