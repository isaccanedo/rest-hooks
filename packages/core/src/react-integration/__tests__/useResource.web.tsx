import {
  CoolerArticleResource,
  UserResource,
  InvalidIfStaleArticleResource,
  photoShape,
  noEntitiesShape,
  ArticleTimedResource,
} from '__tests__/legacy-3';
import { createEntityMeta } from '__tests__/utils';
import { State, ReadShape } from '@rest-hooks/core';
import { initialState } from '@rest-hooks/core';
import React, { Suspense } from 'react';
import { render } from '@testing-library/react';
import nock from 'nock';

// relative imports to avoid circular dependency in tsconfig references

import { normalize } from '@rest-hooks/normalizr';

import {
  makeRenderRestHook,
  makeCacheProvider,
  mockInitialState,
} from '../../../../test';
import { DispatchContext, StateContext } from '../context';
import { useResource } from '../hooks';
import { payload, users, nested } from '../test-fixtures';

async function testDispatchFetch(
  Component: React.FunctionComponent<any>,
  payloads: any[],
) {
  const dispatch = jest.fn();
  const tree = (
    <DispatchContext.Provider value={dispatch}>
      <Suspense fallback={null}>
        <Component />
      </Suspense>
    </DispatchContext.Provider>
  );
  render(tree);
  expect(dispatch).toHaveBeenCalled();
  expect(dispatch.mock.calls.length).toBe(payloads.length);
  let i = 0;
  for (const call of dispatch.mock.calls) {
    delete call[0]?.meta?.createdAt;
    delete call[0]?.meta?.promise;
    expect(call[0]).toMatchSnapshot();
    const action = call[0];
    const res = await action.payload();
    expect(res).toEqual(payloads[i]);
    i++;
  }
}

function ArticleComponentTester({ invalidIfStale = false }) {
  const resource = invalidIfStale
    ? InvalidIfStaleArticleResource
    : CoolerArticleResource;
  const article = useResource(resource.detailShape(), {
    id: payload.id,
  });
  return (
    <div>
      <h3>{article.title}</h3>
      <p>{article.content}</p>
    </div>
  );
}

describe('useResource()', () => {
  let renderRestHook: ReturnType<typeof makeRenderRestHook>;
  const fbmock = jest.fn();

  async function testMalformedResponse(
    payload: any,
    fetchShape: ReadShape<any> = CoolerArticleResource.detailShape(),
  ) {
    nock(/.*/)
      .persist()
      .defaultReplyHeaders({
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Headers': '*',
        'Content-Type': 'application/json',
      })
      .get(`/article-cooler/400`)
      .reply(200, payload);

    const { result, waitForNextUpdate } = renderRestHook(() => {
      return useResource(fetchShape, {
        id: 400,
      });
    });
    expect(result.current).toBeUndefined();
    await waitForNextUpdate();
    expect(result.error).toBeDefined();
    expect((result.error as any).status).toBeGreaterThan(399);
    expect(result.error).toMatchSnapshot();
  }

  function Fallback() {
    fbmock();
    return null;
  }

  beforeAll(() => {
    nock(/.*/)
      .persist()
      .defaultReplyHeaders({
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Headers': 'Access-Token',
        'Content-Type': 'application/json',
      })
      .options(/.*/)
      .reply(200)
      .get(`/article-cooler/${payload.id}`)
      .reply(200, payload)
      .get(`/article-time/${payload.id}`)
      .reply(200, { ...payload, createdAt: '2020-06-07T02:00:15+0000' })
      .delete(`/article-cooler/${payload.id}`)
      .reply(204, '')
      .delete(`/article/${payload.id}`)
      .reply(200, {})
      .get(`/article-cooler/0`)
      .reply(403, {})
      .get(`/article-cooler/666`)
      .reply(200, '')
      .get(`/article-cooler/`)
      .reply(200, nested)
      .get(`/user/`)
      .reply(200, users);
  });

  afterAll(() => {
    nock.cleanAll();
  });

  beforeEach(() => {
    renderRestHook = makeRenderRestHook(makeCacheProvider);
  });

  it('should dispatch an action that fetches', async () => {
    await testDispatchFetch(ArticleComponentTester, [payload]);
  });

  it('should dispatch fetch when sent multiple arguments', async () => {
    function MultiResourceTester() {
      const [article, user] = useResource(
        [
          CoolerArticleResource.detailShape(),
          {
            id: payload.id,
          },
        ],
        [UserResource.listShape(), {}],
      );
      return null;
    }
    await testDispatchFetch(MultiResourceTester, [payload, users]);
  });
  it('should throw same promise until both resolve', async () => {
    const renderRestHook = makeRenderRestHook(makeCacheProvider);
    jest.useFakeTimers();
    nock('http://test.com')
      .get(`/article-cooler/${payload.id}`)
      .delay(1000)
      .reply(200, payload);
    nock('http://test.com').get(`/user/`).delay(2000).reply(200, users);

    function MultiResourceTester() {
      try {
        const [article, user] = useResource(
          [
            CoolerArticleResource.detailShape(),
            {
              id: payload.id,
            },
          ],
          [UserResource.listShape(), {}],
        );
        return article;
      } catch (e: any) {
        // TODO: we're not handling suspense properly so react complains
        // When upgrading test util we should be able to fix this as we'll suspense ourselves.
        if (typeof e.then === 'function') {
          return e;
        } else {
          throw e;
        }
      }
    }
    const { rerender, result } = renderRestHook(MultiResourceTester);
    const firstPromise = result.current;
    expect(firstPromise).toBeDefined();
    expect(typeof firstPromise.then).toBe('function');

    jest.advanceTimersByTime(50);
    rerender();
    expect(result.current).toBe(firstPromise);
    jest.advanceTimersByTime(1000);
    rerender();
    expect(result.current).toBe(firstPromise);
    jest.advanceTimersByTime(2000);
    rerender();
    expect(result.current).toBe(firstPromise);

    // TODO: we're not handling suspense properly so react complains
    // When upgrading test util we should be able to fix this as we'll suspense ourselves.
    const oldError = console.error;
    console.error = () => {};
    jest.runAllTimers();
    await result.current;
    rerender();
    expect(result.current).toMatchInlineSnapshot(`
      CoolerArticleResource {
        "author": null,
        "content": "whatever",
        "id": 5,
        "tags": [
          "a",
          "best",
          "react",
        ],
        "title": "hi ho",
      }
    `);
    // eslint-disable-next-line require-atomic-updates
    console.error = oldError;
    jest.useRealTimers();
  });
  it('should NOT suspend if result already in cache and options.invalidIfStale is false', () => {
    const state: State<unknown> = mockInitialState([
      {
        request: CoolerArticleResource.detailShape(),
        params: payload,
        result: payload,
      },
    ]) as any;

    const tree = (
      <StateContext.Provider value={state}>
        <Suspense fallback={<Fallback />}>
          <ArticleComponentTester />
        </Suspense>
      </StateContext.Provider>
    );
    const { getByText } = render(tree);
    expect(fbmock).not.toBeCalled();
    const title = getByText(payload.title);
    expect(title).toBeDefined();
    expect(title.tagName).toBe('H3');
  });
  it('should NOT suspend even when result is stale and options.invalidIfStale is false', () => {
    const { entities, result } = normalize(payload, CoolerArticleResource);
    const fetchKey = CoolerArticleResource.detailShape().getFetchKey(payload);
    const state = {
      ...initialState,
      entities,
      entityMeta: createEntityMeta(entities),
      results: {
        [fetchKey]: result,
      },
      meta: {
        [fetchKey]: {
          date: 0,
          expiresAt: 0,
        },
      },
    };

    const tree = (
      <StateContext.Provider value={state}>
        <DispatchContext.Provider value={() => Promise.resolve()}>
          <Suspense fallback={<Fallback />}>
            <ArticleComponentTester />
          </Suspense>{' '}
        </DispatchContext.Provider>
      </StateContext.Provider>
    );
    const { getByText } = render(tree);
    expect(fbmock).not.toBeCalled();
    const title = getByText(payload.title);
    expect(title).toBeDefined();
    expect(title.tagName).toBe('H3');
  });
  it('should NOT suspend if result is not stale and options.invalidIfStale is true', () => {
    const { entities, result } = normalize(
      payload,
      InvalidIfStaleArticleResource,
    );
    const fetchKey =
      InvalidIfStaleArticleResource.detailShape().getFetchKey(payload);

    const state = {
      ...initialState,
      entities,
      results: {
        [fetchKey]: result,
      },
      entityMeta: createEntityMeta(entities),
      meta: {
        [fetchKey]: {
          date: Infinity,
          expiresAt: Infinity,
        },
      },
    };

    const tree = (
      <StateContext.Provider value={state}>
        <Suspense fallback={<Fallback />}>
          <ArticleComponentTester invalidIfStale />
        </Suspense>
      </StateContext.Provider>
    );
    const { getByText } = render(tree);
    expect(fbmock).not.toBeCalled();
    const title = getByText(payload.title);
    expect(title).toBeDefined();
    expect(title.tagName).toBe('H3');
  });
  it('should suspend if result stale in cache and options.invalidIfStale is true', () => {
    const { entities, result } = normalize(
      payload,
      InvalidIfStaleArticleResource,
    );
    const fetchKey =
      InvalidIfStaleArticleResource.detailShape().getFetchKey(payload);
    const state = {
      ...initialState,
      entities,
      results: {
        [fetchKey]: result,
      },
      entityMeta: createEntityMeta(entities),
      meta: {
        [fetchKey]: {
          date: 0,
          expiresAt: 0,
        },
      },
    };

    const tree = (
      <StateContext.Provider value={state}>
        <DispatchContext.Provider value={() => Promise.resolve()}>
          <Suspense fallback={<Fallback />}>
            <ArticleComponentTester invalidIfStale />
          </Suspense>
        </DispatchContext.Provider>
      </StateContext.Provider>
    );
    render(tree);
    expect(fbmock).toHaveBeenCalled();
  });

  // taken from integration
  it('should throw errors on bad network', async () => {
    const { result, waitForNextUpdate } = renderRestHook(() => {
      return useResource(CoolerArticleResource.detailShape(), {
        id: '0',
      });
    });
    expect(result.current).toBeUndefined();
    await waitForNextUpdate();
    expect(result.error).toBeDefined();
    expect((result.error as any).status).toBe(403);
  });

  it('should throw errors on bad network (multiarg)', async () => {
    const { result, waitForNextUpdate } = renderRestHook(() => {
      return useResource([
        CoolerArticleResource.detailShape(),
        {
          id: '0',
        },
      ]);
    });
    expect(result.current).toBeUndefined();
    await waitForNextUpdate();
    expect(result.error).toBeDefined();
    expect((result.error as any).status).toBe(403);
  });

  it('should suspend when already has a network error', async () => {
    const error: any = { message: 'network error', status: 400 };
    const FS = { ...CoolerArticleResource.detailShape() };
    FS.options = { ...FS.options, errorExpiryLength: -100 };
    const { result, waitForNextUpdate } = renderRestHook(
      () => {
        return useResource(CoolerArticleResource.detailShape(), {
          id: '0',
        });
      },
      {
        results: [
          {
            request: FS,
            params: {
              id: '0',
            },
            result: error,
            error: true,
          },
        ],
      },
    );
    expect(result.current).toBeUndefined();
    expect(result.error).toBeUndefined();
    await waitForNextUpdate();
    expect(result.error).toBeDefined();
    expect((result.error as any).status).toBe(403);
  });

  it('should suspend when already has a network error (multiarg)', async () => {
    const error: any = { message: 'network error', status: 400 };
    const expiredShape = { ...CoolerArticleResource.detailShape() };
    expiredShape.options = { ...expiredShape.options, errorExpiryLength: -100 };
    const { result, waitForNextUpdate } = renderRestHook(
      () => {
        return useResource([
          CoolerArticleResource.detailShape(),
          {
            id: '0',
          },
        ]);
      },
      {
        results: [
          {
            request: expiredShape,
            params: {
              id: '0',
            },
            result: error,
            error: true,
          },
        ],
      },
    );
    expect(result.current).toBeUndefined();
    expect(result.error).toBeUndefined();
    await waitForNextUpdate();
    expect(result.error).toBeDefined();
    expect((result.error as any).status).toBe(403);
  });

  // Disabled due to: const ready = !!cacheResults || found;
  // Since we will not not error if we have any cacheResults, it's not only not possible to trigger
  // this loop - it required contrived state that cannot be naturally arrived at

  // We need to ensure we don't suspend when error is based on network response
  // as we don't expect another network call to fix it.
  // (fixes infinite loop with badly defined schemas)
  /*it('should throw error when response is bad (on mount)', async () => {
    nock(/.*)
      .persist()
      .defaultReplyHeaders({
        'Access-Control-Allow-Origin': '*',
        'Content-Type': 'application/json',
      })
      .get(`/article-cooler/4000`)
      .reply(200, { data: null });

    const shape = {
      ...CoolerArticleResource.detailShape(),
      schema: { data: CoolerArticleResource },
    };
    const expiredShape = { ...shape };
    expiredShape.options = { ...expiredShape.options, dataExpiryLength: -100 };

    const { result, rerender } = renderRestHook(
      () => {
        return useResource(shape, {
          id: '4000',
        });
      },
      {
        results: [
          {
            request: expiredShape,
            params: {
              id: '4000',
            },
            result: { data: null },
          },
        ],
      },
    );

    const firsterror = result.error;
    expect(result.error).not.toBe(null);
    expect((result.error as any).status).toBe(400);
    expect(result.error).toMatchSnapshot();
    rerender();
    expect(result.error).not.toBe(null);
    expect((result.error as any).status).toBe(400);
    expect(result.error).not.toBe(firsterror);
  });*/

  it('should throw error when response is array when expecting entity', async () => {
    await testMalformedResponse([]);
  });

  it('should throw error when response is {} when expecting entity', async () => {
    await testMalformedResponse({});
  });

  it('should throw error when response is number when expecting entity', async () => {
    await testMalformedResponse(5);
  });

  it('should throw error when response is string when expecting entity', async () => {
    await testMalformedResponse('hi');
  });

  /* TODO: Add these back when we have opt-in required
  it('should throw error when response is string when expecting nested entity', async () => {
    const shape = {
      ...CoolerArticleResource.detailShape(),
      schema: { data: CoolerArticleResource },
    };
    await testMalformedResponse('hi', shape);
  });

  it('should throw error when response is nested string when expecting nested entity', async () => {
    const shape = {
      ...CoolerArticleResource.detailShape(),
      schema: { data: CoolerArticleResource },
    };
    await testMalformedResponse({ data: 5, parcel: 2 }, shape);
  });

  it('should throw error when response is nested missing id when expecting nested entity', async () => {
    const shape = {
      ...CoolerArticleResource.detailShape(),
      schema: { data: CoolerArticleResource },
    };
    await testMalformedResponse(
      { data: { ...payload, id: undefined }, parcel: 2 },
      shape,
    );
  });

  it('should throw error when response is expected Resource inside Record', async () => {
    class Scheme extends SimpleRecord {
      data: CoolerArticleResource = CoolerArticleResource.fromJS();
      optional: UserResource | null = null;
      static schema = {
        data: CoolerArticleResource,
        optional: UserResource,
      };
    }
    const shape = {
      ...CoolerArticleResource.detailShape(),
      schema: Scheme,
    };
    await testMalformedResponse({ data: null }, shape);
  });*/

  it('should resolve parallel useResource() request', async () => {
    const { result, waitForNextUpdate } = renderRestHook(() => {
      return useResource(
        [
          CoolerArticleResource.detailShape(),
          {
            id: payload.id,
          },
        ],
        [UserResource.listShape(), {}],
      );
    });
    await waitForNextUpdate();
    const [article, users] = result.current;
    expect(article instanceof CoolerArticleResource).toBe(true);
    expect(article.title).toBe(payload.title);
    expect(users).toBeDefined();
    expect(users.length).toBeTruthy();
    expect(users[0] instanceof UserResource).toBe(true);
  });

  it('should not suspend with no params to useResource()', () => {
    let article: any;
    const { result } = renderRestHook(() => {
      article = useResource(CoolerArticleResource.detailShape(), null);
      return 'done';
    });
    expect(result.current).toBe('done');
    expect(article).toBeUndefined();
  });

  it('should work with shapes with no entities', async () => {
    const userId = '5';
    const response = { firstThing: '', someItems: [{ a: 5 }] };
    nock(/.*/)
      .defaultReplyHeaders({ 'access-control-allow-origin': '*' })
      .get(`/users/${userId}/simple`)
      .reply(200, response);
    const { result, waitForNextUpdate } = renderRestHook(() => {
      return useResource(noEntitiesShape, { userId });
    });
    // undefined means it threw
    expect(result.current).toBeUndefined();
    await waitForNextUpdate();
    expect(result.current).toStrictEqual(response);
  });

  it('should work with ArrayBuffer shapes', async () => {
    const userId = '5';
    const response = new ArrayBuffer(10);
    nock(/.*/)
      .defaultReplyHeaders({ 'access-control-allow-origin': '*' })
      .get(`/users/${userId}/photo`)
      .reply(200, response);
    const { result, waitForNextUpdate } = renderRestHook(() => {
      return useResource(photoShape, { userId });
    });
    // undefined means it threw
    expect(result.current).toBeUndefined();
    await waitForNextUpdate();
    expect(result.current).toEqual(response);
  });

  it('should work with Serializable shapes', async () => {
    const { result, waitForNextUpdate } = renderRestHook(() => {
      return useResource(ArticleTimedResource.detailShape(), payload);
    });
    // undefined means it threw
    expect(result.current).toBeUndefined();
    await waitForNextUpdate();
    expect(result.current.createdAt.getDate()).toBe(
      result.current.createdAt.getDate(),
    );
    expect(result.current.createdAt).toEqual(
      new Date('2020-06-07T02:00:15+0000'),
    );
    expect(result.current.id).toEqual(payload.id);
    expect(result.current).toBeInstanceOf(ArticleTimedResource);
  });
});
