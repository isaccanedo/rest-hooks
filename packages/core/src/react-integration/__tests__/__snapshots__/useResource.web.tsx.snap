// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useResource() should dispatch an action that fetches 1`] = `
{
  "meta": {
    "args": [
      {
        "id": 5,
      },
      undefined,
    ],
    "key": "GET http://test.com/article-cooler/5",
    "options": {
      "errorPolicy": [Function],
      "fetchInit": {},
      "getFetchInit": [Function],
      "method": "GET",
      "signal": undefined,
      "url": [Function],
      "useFetchInit": [Function],
    },
    "reject": [Function],
    "resolve": [Function],
    "schema": [Function],
    "throttle": true,
    "type": "read",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;

exports[`useResource() should dispatch fetch when sent multiple arguments 1`] = `
{
  "meta": {
    "args": [
      {
        "id": 5,
      },
      undefined,
    ],
    "key": "GET http://test.com/article-cooler/5",
    "options": {
      "errorPolicy": [Function],
      "fetchInit": {},
      "getFetchInit": [Function],
      "method": "GET",
      "signal": undefined,
      "url": [Function],
      "useFetchInit": [Function],
    },
    "reject": [Function],
    "resolve": [Function],
    "schema": [Function],
    "throttle": true,
    "type": "read",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;

exports[`useResource() should dispatch fetch when sent multiple arguments 2`] = `
{
  "meta": {
    "args": [
      {},
      undefined,
    ],
    "key": "GET http://test.com/user/",
    "options": {
      "errorPolicy": [Function],
      "fetchInit": {},
      "getFetchInit": [Function],
      "method": "GET",
      "signal": undefined,
      "url": [Function],
      "useFetchInit": [Function],
    },
    "reject": [Function],
    "resolve": [Function],
    "schema": [
      [Function],
    ],
    "throttle": true,
    "type": "read",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;

exports[`useResource() should throw error when response is {} when expecting entity 1`] = `
[Error: Error processing GET http://test.com/article-cooler/400

Full Schema: {
  "name": "CoolerArticleResource",
  "schema": {
    "author": {
      "name": "UserResource",
      "schema": {},
      "key": "http://test.com/user/"
    }
  },
  "key": "http://test.com/article-cooler/"
}

Error:
Attempted to initialize CoolerArticleResource with an array, but named members were expected

This is likely due to a malformed response.
Try inspecting the network response or fetch() return value.
Or use debugging tools: https://resthooks.io/docs/guides/debugging
Learn more about schemas: https://resthooks.io/docs/api/schema
If this is a mistake, you can disable this check by setting static automaticValidation = 'silent'

First three members: []]
`;

exports[`useResource() should throw error when response is array when expecting entity 1`] = `
[Error: Error processing GET http://test.com/article-cooler/400

Full Schema: {
  "name": "CoolerArticleResource",
  "schema": {
    "author": {
      "name": "UserResource",
      "schema": {},
      "key": "http://test.com/user/"
    }
  },
  "key": "http://test.com/article-cooler/"
}

Error:
Attempted to initialize CoolerArticleResource with an array, but named members were expected

This is likely due to a malformed response.
Try inspecting the network response or fetch() return value.
Or use debugging tools: https://resthooks.io/docs/guides/debugging
Learn more about schemas: https://resthooks.io/docs/api/schema
If this is a mistake, you can disable this check by setting static automaticValidation = 'silent'

First three members: []]
`;

exports[`useResource() should throw error when response is number when expecting entity 1`] = `
[Error: Error processing GET http://test.com/article-cooler/400

Full Schema: {
  "name": "CoolerArticleResource",
  "schema": {
    "author": {
      "name": "UserResource",
      "schema": {},
      "key": "http://test.com/user/"
    }
  },
  "key": "http://test.com/article-cooler/"
}

Error:
Attempted to initialize CoolerArticleResource with an array, but named members were expected

This is likely due to a malformed response.
Try inspecting the network response or fetch() return value.
Or use debugging tools: https://resthooks.io/docs/guides/debugging
Learn more about schemas: https://resthooks.io/docs/api/schema
If this is a mistake, you can disable this check by setting static automaticValidation = 'silent'

First three members: []]
`;

exports[`useResource() should throw error when response is string when expecting entity 1`] = `
[Error: Error processing GET http://test.com/article-cooler/400

Full Schema: {
  "name": "CoolerArticleResource",
  "schema": {
    "author": {
      "name": "UserResource",
      "schema": {},
      "key": "http://test.com/user/"
    }
  },
  "key": "http://test.com/article-cooler/"
}

Error:
Attempted to initialize CoolerArticleResource with an array, but named members were expected

This is likely due to a malformed response.
Try inspecting the network response or fetch() return value.
Or use debugging tools: https://resthooks.io/docs/guides/debugging
Learn more about schemas: https://resthooks.io/docs/api/schema
If this is a mistake, you can disable this check by setting static automaticValidation = 'silent'

First three members: []]
`;
